demos/fleet_provisioning/fleet_provisioning_with_csr/fleet_provisioning_csr_demo.c

```diff
@@ -0,0 +1,1271 @@
+ * length is output here.
+ */
+static bool parseRegisterThingResponse( const char * response,
+                                        size_t length,
+                                        char * thingNameBuffer,
+                                        size_t * thingNameBufferLength );
+
+/**
+ * @brief Run the MQTT process loop to get a response.
+ */
+static bool waitForResponse( void );
+
+/**
+ * @brief Subscribe to the CreateCertificateFromCsr accepted and rejected topics.
+ */
+static bool subscribeToCsrResponseTopics( void );
+
+/**
+ * @brief Unsubscribe from the CreateCertificateFromCsr accepted and rejected topics.
+ */
+static bool unsubscribeFromCsrResponseTopics( void );
+
+/**
+ * @brief Subscribe to the RegisterThing accepted and rejected topics.
+ */
+static bool subscribeToRegisterThingResponseTopics( void );
+
+/**
+ * @brief Unsubscribe from the RegisterThing accepted and rejected topics.
+ */
+static bool unsubscribeFromRegisterThingResponseTopics( void );
+
+/**
+ * @brief Save the certificate to the path specified by #PROVISIONING_CERT_PATH.
+ */
+static bool saveCertificate( const char * certificate,
+                             size_t certificateLength );
+/*-----------------------------------------------------------*/
+
+static void provisioningPublishCallback( MQTTPublishInfo_t * pPublishInfo,
+                                         uint16_t packetIdentifier )
+{
+    FleetProvisioningStatus_t status;
+    FleetProvisioningTopic_t api;
+
+    /* Silence compiler warnings about unused variables. */
+    ( void ) packetIdentifier;
+
+    status = FleetProvisioning_MatchTopic( pPublishInfo->pTopicName,
+                                           pPublishInfo->topicNameLength, &api );
+
+    if( status != FleetProvisioningSuccess )
+    {
+        LogError( ( "Unexpected publish message received. Topic: %.*s.",
+                    ( int ) pPublishInfo->topicNameLength,
+                    ( const char * ) pPublishInfo->pTopicName ) );
+    }
+    else
+    {
+        if( api == FleetProvCborCreateCertFromCsrAccepted )
+        {
+            LogInfo( ( "Fleet Provisioning CreateCertificateFromCsr accepted." ) );
+
+            responseStatus = ResponseAccepted;
+
+            ( void ) memcpy( ( void * ) payloadBuffer,
+                             ( const void * ) pPublishInfo->pPayload,
+                             ( size_t ) pPublishInfo->payloadLength );
+
+            payloadLength = pPublishInfo->payloadLength;
+        }
+        else if( api == FleetProvCborCreateCertFromCsrRejected )
+        {
+            LogError( ( "Fleet Provisioning CreateCertificateFromCsr rejected." ) );
+
+            responseStatus = ResponseRejected;
+        }
+        else if( api == FleetProvCborRegisterThingAccepted )
+        {
+            LogInfo( ( "Fleet Provisioning RegisterThing accepted." ) );
+
+            responseStatus = ResponseAccepted;
+
+            ( void ) memcpy( ( void * ) payloadBuffer,
```

Comment:
Instead of having 2 places of calling `memcpy`, a flag can be set and that can be used later in the function to have a single copy operation logic

Reply:
That seems like it would only make it more complicated